src/components/legend/style.js

```diff
@@ -30,7 +30,7 @@ module.exports = function style(s, gd) {
         if(gd._fullLayout.legend.valign === 'top') valignFactor = 1.0;
         if(gd._fullLayout.legend.valign === 'bottom') valignFactor = -1.0;
         var markerOffsetY = valignFactor * (0.5 * (d[0].lineHeight - d[0].height + 3));
-        if(markerOffsetY) layers.attr('transform', 'translate(0,' + markerOffsetY + ')');
```

Comment:
Hmm. But `'middle'` is the default value isn't it. Why didn't our tests catch sooner?

Reply:
All the mocks were rendered properly because the `transform` attribute was not applied. However, if there's already a `transform` attribute on the svg group because `valign` was `top`, then it needs to be removed on `relayout` if `valign` is switched back to `middle`.

Reply:
> However, if there's already a transform attribute on the svg group because valign was top, then it needs to be removed on relayout if valign is switched back to middle.

OK this makes sense. I don't understand how the `transform` gets removed and why this patch fixes the bug though. `Boolean(NaN)` is `false` so the old and this new line look equivalent to me.

Reply:
Wouldn't something like

```js
var valign = gd._fullLayout.valign;

if(valign === 'middle') {
  layers.attr('transform', null); // this here is a fun d3 trick to unset DOM attributes
} else {
  var factor = {top: 1, bottom: -1}[valign];
  var markerOffsetY = factor * (0.5 * (d[0].lineHeight - d[0].height + 3));
  layers.attr('transform', 'translate(0,' + markerOffsetY + ')');
}
```

be what we're looking for?

Reply:
Sorry @etpinard for the unclear comment. The `transform` doesn't get removed but it gets set to `translate(0,0)`.

I think before we had:
`valign: top` -> `transform="translate(0, offsetForTop)"`
`valign: middle` -> no change to the transform attribute :warning: problematic on `relayout`
`valign: top` -> `transform="translate(0, offsetForBottom)"`

Now we have `valign: middle` ->  `transform="translate(0, 0)"`

Reply:
Ah ok I get it. I didn't see the newly-added

```js
var valignFactor = 0;
```

line.

Reply:
When does `markerOffsetY` evaluate to a `NaN` now?

Reply:
Your solution would work except that we should check whether `markerOffsetY` is a number. It might not be when this function is visited and errors would be thrown in the console because of garbage `translate(0, NaN)`. This is why I initially added `if(!isNaN(markerOffsetY))`!

Reply:
I'd have to check to make sure but I think it's called on first dray to figure out how big the legend will be. At this point, `d[0].lineHeight`  `d[0].height` are not defined and we get a `NaN`. Although everything renders correctly in the end, we get the following in the console:
`d3.js:663 Error: <g> attribute transform: Trailing garbage, "translate(0,NaN)".`

Reply:
That is indeed what is happening

Reply:
In commit c1b871d, I reused most of the above :smile: !